apps/studio/src/server/modules/auth/singpass/singpass.router.ts

```diff
@@ -0,0 +1,170 @@
+import set from "lodash/set"
+
+import type { SessionData } from "~/lib/types/session"
+import { DASHBOARD } from "~/lib/routes"
+import {
+  singpassCallbackSchema,
+  singpassLoginSchema,
+  singpassStateSchema,
+} from "~/schemas/auth/singpass"
+import { publicProcedure, router } from "~/server/trpc"
+import { safeSchemaJsonParse } from "~/utils/zod"
+import { logUserEvent } from "../../audit/audit.service"
+import { recordUserLogin } from "../auth.service"
+import { getAuthorizationUrl, login } from "./singpass.service"
+
+export const singpassRouter = router({
+  login: publicProcedure
+    .input(singpassLoginSchema)
+    .mutation(async ({ ctx, input: { landingUrl } }) => {
+      // NOTE: The Singpass login flow is not the first login mechanism that the
+      // user encounters, as they should have completed the email OTP
+      // verification before this. Hence, the user will need to have a partial
+      // user session created before this step.
+      if (!ctx.session.singpass?.sessionState?.userId) {
+        throw new TRPCError({
+          code: "BAD_REQUEST",
+          message: "Email verification has not been completed",
+        })
+      }
+
+      const { userId, verificationToken } = ctx.session.singpass.sessionState
+
+      ctx.logger.info(
+        { landingUrl },
+        `Starting Singpass login flow: ${landingUrl.toString()}`,
+      )
+
+      const { authorizationUrl, session } = getAuthorizationUrl()
+
+      // Reset session state
+      ctx.session.destroy()
+
+      set(ctx.session, "singpass.sessionState", {
+        ...session,
+        userId,
+        verificationToken,
+      })
+
+      await ctx.session.save()
+
+      return {
+        redirectUrl: authorizationUrl,
+      }
+    }),
+
+  callback: publicProcedure
+    .input(singpassCallbackSchema)
+    .query(async ({ ctx, input: { state, code } }) => {
+      if (!ctx.session.singpass?.sessionState) {
+        ctx.logger.warn("No Singpass session state found")
+
+        throw new TRPCError({
+          code: "BAD_REQUEST",
+          message: "Invalid login flow",
+        })
+      }
+
+      const parsedState = safeSchemaJsonParse(singpassStateSchema, state)
+
+      if (!parsedState.success) {
+        ctx.logger.error(
+          { state, error: parsedState.error },
+          "Invalid Singpass callback state",
+        )
+
+        throw new TRPCError({
+          code: "BAD_REQUEST",
+          message: "Invalid Singpass callback state",
+        })
+      }
+
+      const { codeVerifier, nonce, userId, verificationToken } =
+        ctx.session.singpass.sessionState
+      ctx.session.destroy()
+
+      if (!code || !codeVerifier || !nonce || !userId) {
+        ctx.logger.error(
+          { code, codeVerifier, nonce, userId },
+          "Invalid Singpass session state",
+        )
+
+        throw new TRPCError({
+          code: "BAD_REQUEST",
+          message: "Invalid Singpass session state",
+        })
+      }
```

Comment:
question: since these 4 values are defined, im assuming the check here is for runtime situation where these are somehow undefined (correct me if im wrong!). Do we know what kinda situation that might be, or is this just defensive programming?